File: diffusion.py
import argparse
from pathlib import Path
from typing import List
from time import perf_counter_ns
import numpy as np
import scipy as sp
import cupyx.scipy.signal as sgx
import cupy as cp
import torch
import cv2


"""
TODO:
- Profile to verify why numpy is slower (hypothesis: no benefit of underlying C++ data structure since doing single accesses at a time)
"""


def save_results(c, implementation):
     if implementation == 'naive':
          with open(Path('python_' + implementation + '.txt'), 'w') as file:
               for row in c:
                    file.writelines(','.join([str(i) for i in row]) + '\n')
     elif implementation in {'numpy', 'scipy', 'opencv', 'cupy', 'torch'}:
          np.savetxt('python_' + implementation + '.txt', c, fmt='%f', delimiter=',')
     else:
          raise ValueError(f'Implementation {implementation} does\'t exist')


def diffuse_torch(c : List[List[float]],
                 c_tmp : List[List[float]],
                 aux : float,
                 num_steps : int):
     width = len(c)
     gpu = torch.device('cuda')
     c = torch.tensor(c, dtype=torch.float64, device=gpu)
     c = torch.unsqueeze(torch.unsqueeze(c, 0), 0)
     c_tmp = torch.tensor(c_tmp, dtype=torch.float64, device=gpu)
     c_tmp = torch.unsqueeze(torch.unsqueeze(c_tmp, 0), 0)
     kernel = aux * torch.tensor([[0, 1, 0],
                                  [1, -4, 1],
                                  [0, 1, 0]],
                                  dtype=torch.float64,
                                  device=gpu)
     kernel = torch.unsqueeze(torch.unsqueeze(kernel, 0), 0)
     assert c.is_cuda
     assert c_tmp.is_cuda
     assert kernel.is_cuda
     time_start = perf_counter_ns()
     for _ in range(num_steps):
          c_tmp = c + torch.nn.functional.pad(
               torch.nn.functional.conv2d(c, kernel, padding='valid'),
               (1,1,1,1),
               'constant',
               0
          )
          c, c_tmp = c_tmp, c
     time_stop = perf_counter_ns()
     c_reshaped = torch.reshape(c, (width, width)).cpu().numpy()
     return  c_reshaped, time_stop - time_start


def diffuse_cupy(c : List[List[float]],
                 c_tmp : List[List[float]],
                 aux : float,
                 num_steps : int):
     c = cp.array(c)
     c_tmp = cp.array(c_tmp)
     kernel = aux * cp.array([[0, 1, 0],
                              [1, -4, 1],
                              [0, 1, 0]])
     time_start = perf_counter_ns()
     for _ in range(num_steps):
          c_tmp = c + cp.pad(
               sgx.convolve2d(c, kernel, mode='valid'),
                    1,
                    'constant'
               )
          c, c_tmp = c_tmp, c
     time_stop = perf_counter_ns()
     return c, time_stop - time_start


def diffuse_opencv(c : List[List[float]],
                   c_tmp : List[List[float]],
                   aux : float,
                   num_steps : int):
     c = np.array(c)
     c_tmp = np.array(c_tmp)
     kernel = aux * np.array([[0, 1, 0],
                              [1, -3, 1],
                              [0, 1, 0]])
     time_start = perf_counter_ns()
     for _ in range(num_steps):
          c_tmp = cv2.filter2D(src=c, kernel=kernel, borderType=cv2.BORDER_CONSTANT, value=0)
          c, c_tmp = c_tmp, c
     time_stop = perf_counter_ns()
     c = np.pad(c, 1,'constant')
     return c, time_stop - time_start


def diffuse_scipy(c : List[List[float]],
                  c_tmp : List[List[float]],
                  aux : float,
                  num_steps : int):
     c = np.array(c)
     c_tmp = np.array(c_tmp)
     kernel = aux * np.array([[0, 1, 0],
                              [1, -4, 1],
                              [0, 1, 0]])
     time_start = perf_counter_ns()
     for _ in range(num_steps):
          c_tmp = c + np.pad(
               sp.signal.convolve2d(c, kernel, mode='valid'),
                    1,
                    'constant'
               )
          c, c_tmp = c_tmp, c
     time_stop = perf_counter_ns()
     return c, time_stop - time_start


def diffuse_numpy(c : List[List[float]],
                  c_tmp : List[List[float]],
                  aux : float,
                  num_steps : int):
     c = np.array(c)
     c_tmp = np.array(c_tmp)
     time_start = perf_counter_ns()
     for _ in range(num_steps):
          for i in range(1, len(c) - 1):
               for j in range(1, len(c[0]) - 1):
                    c_tmp[i][j] = (c[i][j] + aux *
                                   (c[i - 1][j] + c[i + 1][j] +
                                    c[i][j - 1] + c[i][j + 1] -
                                    4 * c[i][j]))
          c, c_tmp = c_tmp, c
     time_stop = perf_counter_ns()
     return c, time_stop - time_start


def diffuse_naive(c : List[List[float]],
                  c_tmp : List[List[float]],
                  aux : float,
                  num_steps : int):
     time_start = perf_counter_ns()
     for _ in range(num_steps):
          for i in range(1, len(c) - 1):
               for j in range(1, len(c[0]) - 1):
                    c_tmp[i][j] = (c[i][j] + aux *
                                   (c[i - 1][j] + c[i + 1][j] +
                                    c[i][j - 1] + c[i][j + 1] -
                                    4 * c[i][j]))
          c, c_tmp = c_tmp, c
     time_stop = perf_counter_ns()
     return c, time_stop - time_start


def initialize_concentration(L : float, N : float, h : float):
     bound = 0.125 * L
     # +2 to add a border of boundary cells (boundary condition)
     c = [[0.0] * (N+2) for _ in range(N+2)]
     # Initialize non-boundary cells
     for i in range(N):
          for j in range(N):
               if (abs(i * h - 0.5 * L) < bound and
                   abs(j * h - 0.5 * L) < bound):
                    c[i+1][j+1] = 1.0
     c_tmp = [row[:] for row in c]
     return c, c_tmp


def get_constants(D : float, L : float, N : int):
     h = L / (N - 1)          # length between grid elements
     dt = h * h / (4 * D)     # maximum timestamp for numeric stability
     aux = dt * D / (h * h)   # all constant terms in diffusion equation
     return h, dt, aux


if __name__ == '__main__':
     parser = argparse.ArgumentParser()
     parser.add_argument('-D', type=float, required=False, default=1,
                         help='Diffusion constant')
     parser.add_argument('-L', type=float, required=False, default = 2,
                         help='Domain size')
     parser.add_argument('-N', type=int, required=False, default=100,
                         help='Number of grid points along each dim')
     parser.add_argument('-T', type=float, required=False, default=1.0,
                         help='Simulation time in seconds')
     parser.add_argument('--implementation', type=str, required=False, default='naive',
                         help='Diffusion implementation to execute')
     parser.add_argument('--output', action='store_true', required=False,
                         help='Whether to store results')
     args = parser.parse_args()
     D = args.D
     L = args.L
     N = args.N
     T = args.T
     implementation = args.implementation
     output = args.output
     h, dt, aux = get_constants(D, L, N)
     num_steps = int(T / dt) + 1
     c, c_tmp = initialize_concentration(L, N, h)
     if implementation == 'naive':
          c, time_ns = diffuse_naive(c, c_tmp, aux, num_steps)
     elif implementation == 'numpy':
          c, time_ns = diffuse_numpy(c, c_tmp, aux, num_steps)
     elif implementation == 'scipy':
          c, time_ns = diffuse_scipy(c, c_tmp, aux, num_steps)
     elif implementation == 'opencv':
          c, time_ns = diffuse_torch(c, c_tmp, aux, num_steps)
     elif implementation == 'cupy':
          c, time_ns = diffuse_cupy(c, c_tmp, aux, num_steps)
     elif implementation == 'torch':
          c, time_ns = diffuse_torch(c, c_tmp, aux, num_steps)
     else:
          raise ValueError(f'Implementation {implementation} does\'t exist')
     if output:
          save_results(c, implementation)
     print(f'Elapsed time: {time_ns}ns')

File: diffusion.cpp
#include <iostream>
#include <chrono>
#include <cmath>
#include <fstream>
#include <vector>
#include <omp.h>
#include "diffusioncuda.h"


/*
* TODO:
*   - Find out why std::abs() works but abs() doesn't in initialization
*
* Compile with:
* g++ -O3 -Wall -Wextra -pedantic -Wall -Wconversion -Wextra -pedantic -std=c++14 -fopenmp -o diffusion diffusion.cpp
*/


void save_results(const std::vector<double> &c,
                    const std::string &implementation) {
    std::string filename = "cpp_" + implementation + ".txt";
    // M = N + 2, the length with padding
    size_t M = (size_t) sqrt(c.size());
    std::ofstream ofs(filename, std::ofstream::out);
    for (size_t i = 0; i < M; ++i) {
        for (size_t j = 0; j < M; ++j) {
            ofs << c[i * M + j];
            if (j == M - 1) {
                ofs << "\n";
            } else {
                ofs << ",";
            }
        }
    }
    ofs.close();
}


void vector_to_arr(std::vector<double> &c,
                    double *arr,
                    const size_t M) {
    for (size_t i = 0; i < M; ++i) {
        for (size_t j = 0; j < M; ++j) {
            arr[i * M + j] = c[i * M + j];
        }
    }
}


void arr_to_vector(std::vector<double> &c,
                    double *arr,
                    const size_t M) {
    for (size_t i = 0; i < M; ++i) {
        for (size_t j = 0; j < M; ++j) {
            c[i * M + j] = arr[i * M + j];
        }
    }
}


std::chrono::nanoseconds diffuse_openmp(std::vector<double> &c,
                            std::vector<double> &c_tmp,
                            const double aux,
                            const size_t num_steps) {
    // M = N + 2, the length with padding
    const size_t M = (size_t) sqrt(c.size());
    auto time_start = std::chrono::steady_clock::now();
    for (size_t step = 0; step < num_steps; ++step) {
#pragma omp parallel for collapse(2)
        for (size_t i = 1; i < M - 1; ++i) {
            for (size_t j = 1; j < M - 1; ++j) {
                c_tmp[i * M + j] = c[i * M + j] + aux * (
                    c[i * M + (j + 1)] + c[i * M + (j - 1)] +
                    c[(i + 1) * M + j] + c[(i - 1) * M + j] -
                    4 * c[i * M + j]);
            }
        }
        std::swap(c, c_tmp);
    }
    auto time_stop = std::chrono::steady_clock::now();
    auto time_elapsed = std::chrono::duration_cast<std::chrono::nanoseconds>(time_stop - time_start);
    return time_elapsed;
}


__inline__ void step_const_c(const std::vector<double> &__restrict c,
                    std::vector<double> &c_tmp,
                    const size_t &M,
                    const double &aux) {
    for (size_t i = 1; i < M - 1; ++i) {
        for (size_t j = 1; j < M - 1; ++j) {
            c_tmp[i * M + j] = c[i * M + j] + aux * (
                c[i * M + (j + 1)] + c[i * M + (j - 1)] +
                c[(i + 1) * M + j] + c[(i - 1) * M + j] -
                4 * c[i * M + j]);
        }
    }
}


std::chrono::nanoseconds diffuse_const_c(std::vector<double> &c,
                            std::vector<double> &c_tmp,
                            const double aux,
                            const size_t num_steps) {
    // M = N + 2, the length with padding
    const size_t M = (size_t) sqrt(c.size());
    auto time_start = std::chrono::high_resolution_clock::now();
    for (size_t step = 0; step < num_steps; ++step) {
        step_const_c(c, c_tmp, M, aux);
        std::swap(c, c_tmp);
    }
    auto time_stop = std::chrono::high_resolution_clock::now();
    auto time_elapsed = std::chrono::duration_cast<std::chrono::nanoseconds>(time_stop - time_start);
    return time_elapsed;
}


std::chrono::nanoseconds diffuse_naive(std::vector<double> &c,
                            std::vector<double> &c_tmp,
                            const double aux,
                            const size_t num_steps) {
    // M = N + 2, the length with padding
    const size_t M = (size_t) sqrt(c.size());
    auto time_start = std::chrono::steady_clock::now();
    for (size_t step = 0; step < num_steps; ++step) {
        for (size_t i = 1; i < M - 1; ++i) {
            for (size_t j = 1; j < M - 1; ++j) {
                c_tmp[i * M + j] = c[i * M + j] + aux * (
                    c[i * M + (j + 1)] + c[i * M + (j - 1)] +
                    c[(i + 1) * M + j] + c[(i - 1) * M + j] -
                    4 * c[i * M + j]);
            }
        }
        std::swap(c, c_tmp);
    }
    auto time_stop = std::chrono::steady_clock::now();
    auto time_elapsed = std::chrono::duration_cast<std::chrono::nanoseconds>(time_stop - time_start);
    return time_elapsed;
}


void initialize_concentration(std::vector<double> &c,
                                const double &L,
                                const size_t &N,
                                const double &h) {
    const double bound = 0.125 * L;
    for (size_t i = 0; i < N; ++i) {
        for (size_t j = 0; j < N; ++j) {
            if (std::abs(i * h - 0.5 * L) < bound &&
                std::abs(j * h - 0.5 * L) < bound) {
                c[(i + 1) * (N + 2) + (j + 1)] = 1.0;
            }
        }
    }
}


int main(int argc, char *argv[]) {
    if (argc < 7) {
        fprintf(stderr, "Usage: %s D L N T out implementation\n", argv[0]);
        return 1;
    }

    const double D = std::stod(argv[1]);        // diffusion constant
    const double L = std::stod(argv[2]);        // domain size
    const size_t N = std::stoul(argv[3]);       // number of grid points along each dim
    const double T = std::stod(argv[4]);        // simulation time in seconds
    const size_t output = std::stoul(argv[5]);  // whether to store results, yes == 1
    const std::string implementation = argv[6]; // version of program to run

    const double h = L / (N - 1);               // length between grid elements
    const double dt = h * h / (4 * D);          // maximum timestamp for numeric stability
    const double aux = dt * D / (h * h);        // all constant terms in diffusion equation
    const size_t num_steps = (size_t) ((T / dt) + 1);

    std::vector<double> c((N+2) * (N+2), 0.0);
    std::vector<double> c_tmp((N+2) * (N+2), 0.0);
    initialize_concentration(c, L, N, h);
    std::chrono::nanoseconds time_elapsed;
    if (implementation.compare("naive") == 0) {
        time_elapsed = diffuse_naive(c, c_tmp, aux, num_steps);
    } else if (implementation.compare("const") == 0) {
        time_elapsed = diffuse_const_c(c, c_tmp, aux, num_steps);
    } else if (implementation.compare("openmp") == 0) {
        time_elapsed = diffuse_openmp(c, c_tmp, aux, num_steps);
    } else if (implementation.compare("cuda") == 0) {
        double *arr = (double *) malloc((N + 2) * (N + 2) * sizeof(double));
        vector_to_arr(c, arr, N + 2);
        time_elapsed = diffuse_cuda(arr, aux, N + 2, num_steps);
        arr_to_vector(c, arr, N + 2);
    }  else {
        std::cout << "Implementation " << implementation << " not found\n";
        exit(1);
    }
    printf("Elapsed time: %luns\n", time_elapsed.count());
    if (output == 1) {
        save_results(c, implementation);
    }
}
